docs: Add release notes for Metals 1.0.0 (#5453)
scalameta/metals: website/blog/2023-07-19-silver.md

---
author: Tomasz Godzik
title: Metals v1.0.0 - Silver
authorURL: https://twitter.com/TomekGodzik
authorImageURL: https://github.com/tgodzik.png
---

We're extremely happy and proud to finally announce the release of Metals
v1.0.0, which is long overdue. This release is a culmination of a lot of work

> from the community and the core Metals team. As it stands Metals supports most

Reviewer: ```suggestion
from the community and the core Metals team. As it stands, Metals supports most
```

of the important LSP features and beyond, which makes it equivalent to many
IDEs, but in your favorite editor.

Bringing in version 1.0.0 doesn't mean that we don't have anything else to
improve, but it means that we are confident that Metals is ready for production
use in business and personal projects. We will continue to improve Metals and
add new features, but we will also fully focus on making Metals the most
productive and stable environment for all Scala developers. For that goal we are

> are providing a number of reporting improvements that can help identify any

Reviewer: ```suggestion
providing a number of reporting improvements that can help identify any
```

> problems you might have and there is more of those coming your way.

Reviewer: ```suggestion
problems you might have and there are more of those coming your way.
```

> Note for plugin authors: Starting from this release any changes in minor version

Reviewer: ```suggestion
Note for plugin authors: Starting from this release, any changes in minor version
```

  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center"></td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">5</td>
  </tr>
</tbody>
</table>

For full details: [https://github.com/scalameta/metals/milestone/_num_?closed=1]
(https://github.com/scalameta/metals/milestone/_num_?closed=1)

Metals is a language server for Scala that works with VS Code, Vim, Emacs and
Sublime Text. Metals is developed at the [Scala Center](https://scala.epfl.ch/)
and [VirtusLab](https://virtuslab.com) with the help from
[Lunatech](https://lunatech.com) along with contributors from the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- [Add multi-root projects support](#add-multi-root-projects-support)
- [Hover support in Java files](#hover-support-in-java-files)
- [Command for running specified scalafix rules](#command-for-running-specified-scalafix-rules)
- [Add clean-all command](#add-clean-all-command)
- [Better support for Scala CLI](#better-support-for-scala-cli)
- [Support new Scala versions](#support-new-scala-versions)
- [Miscellaneous](#miscellaneous)

## Add multi-root projects support

> Up until this release Metals could only be used in workspaces with a single

Reviewer: ```suggestion
Up until this release, Metals could only be used in workspaces with a single
```

> root. This means that if for example you used the `Add folder` option in Visual

Reviewer 2: The full name is actually: "Add Folder to Workspace..."

the doctor interface). Test explorer will also show all tests from all folders
clearly marked with the folder name. Most other commands will be invoked in the
currently active folder in order not to create large cascading computations such
as multiple folders compiling everything.

This tremendously useful improvement was authored by
[kasiaMarek](https://github.com/kasiaMarek). Take a look at the [blogpost]( TODO
link-after-release) for more details!

## Hover support in Java files

Another improvement for better experience in Java files was contributed by
[VankaTaganai](https://github.com/VankaTaganai). Starting from this release you
will be able to hover on names in Java files to see more information about them.
This feature is primarily aimed at Scala developers that need to from time to
time look into Java source and figure out what's going on there. This will not
provide an experience of a full Java language server, but it will provide some
basic functionality that will make it easier to work with Java sources.

TODO Gif

Please let us know about any issues! This is a new feature and we expect that it
might take some time to get it ironed out.

> 

Author: Hmm... Do we still need any flags though? I think @Arthurm1 removed the parts of code that would require it, https://github.com/scalameta/metals/pull/5375

> 

Reviewer 3: Oooo, I totally missed this. So then the logic to add extra flags to start Metals is no longer needed at all?

> 

Author: Let me check that.

> 

Author: Yep, looks like it works without the flags.

> 

Reviewer 3: Ah ok cool. We should revert https://github.com/scalameta/metals-vscode/pull/1395 as well then.

> 

Author: Done! https://github.com/scalameta/metals-vscode/pull/1404

## Command for running specified scalafix rules

Users will now be able to run a single Scalafix rule on a file in your project.
This might be particularly useful if you don't want to run everything at once
and see exactly what is being done by each Scalafix rule.

This great usability improvement was done by
[LaurenceWarne](https://github.com/LaurenceWarne).

## Add clean-all command

This new feature has been demanded by users for a long while. Thanks to work
done by [jkciesluk](https://github.com/jkciesluk), you will be able to run

> `metals.reset-workspace` command to remove all the possible caches and compile

Reviewer 3: ```suggestion
the `metals.reset-workspace` command to remove all the possible caches and compilation
```

> artifact that might be causing issues in your workspace.

Reviewer 3: ```suggestion
artifacts that might be causing issues in your workspace.
```

> This should be seen as a last resort and we hope that anyway if you keep having
> issues you will report them to us so we can properly fix them.

Reviewer 3: ```suggestion
This should be seen as a last resort and used in as an alternative to the somewhat popular practice of blowing away your `.metals/` and `.bloop/` directories. Please do keep in mind that if you're finding that you need this often, it's probably a bug and should be reported.
```

Author: Thanks! That sounds better!

## Better support for Scala CLI

With Scala CLI becoming the official runner for Scala 3, we have been working on
improving the support for it in Metals. This release brings a number of
improvements that aim to improve both Scala CLI experience and the overall Scala
experience.

This includes the following:

- If there are no build tools present Metals will now fallback to Scala CLI,

>   which means that even if a users creates a single file they will be able to

Reviewer: ```suggestion
  which means that even if a user creates a single file they will be able to
```

Here we only show a short summary of the main feature from the release notes.

#### Linting

Scala 3.3.0 brings back linting to the Scala compiler. Right now, you can enable
checking for unused symbols and discarded values. More linting options will come
soon in the following Scala 3.3.x releases.

#### Fewer braces

Braces around method parameters can now be replaced with a colon. This can lead
to cleaner, shorter, and often more readable code in places like configuration
DSLs or test case definitions.

#### Boundary and break

Two new methods were added to the standard library: boundary and break. They are
safer and more expressive replacements for non-local returns, which were
deprecated recently.

break allows for a type-safe early escape from anywhere inside the block
delimited by boundary to its end, returning the passed value from the entire
block.

#### The new default implementation of lazy vals

Last but not least, we have changed the default implementation of lazy vals. The
new implementation has better performance and is safer under parallel access.

> This may result in improvements in the performance of effect systems.

Reviewer 3: No strong feeling on this, but I don't actually think this is needed. It might even be confusing that we're outlining features of the Scala release here in the Metals release. I'd focus on what this release of Metals brings you and link to the relevant places elsewhere. I'm not sure this add a lot to the Metals release notes.
